### PR TITLE
Bug 3443

### PR DESCRIPTION
**Pull request by skulle**
---
name: Work in progres, bug-3443
about: Enable filtering via root elements
---

## Motivation and Context

#3443 

## How Has This Been Tested?
Currently i am testing locally with a dita sample map only

## Type of Changes
For map-first pre-processing:
1. map-branch-filter step: Add an additional step to the FilterUtils.needExclude method to check if a filtering attribute exists at the referenced map/topic. Remove filtered maps/topics from the job's file section

## Checklist

- My code follows the code style of this project.
    -  <https://github.com/dita-ot/dita-ot/wiki/Java-Coding-Conventions>
    -  <https://github.com/dita-ot/dita-ot/wiki/XSLT-Coding-Conventions>
- I have updated the unit tests to reflect the changes in my code.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,14 @@
+#Eclipse
```
Comment (jelovirt) [inline .gitignore]: You can ignore these in your local `.git/info/exclude`, you don't need to add these to `.gitignore`

Comment (skulle) [reply .gitignore]: That is practical, i didn't know that feature. I will revert the gitignore changes...

**Comment by jelovirt** [comment]
If the issue is supporting filter cases where the whole topic content is being filtered out, why do we need to changed key resolution code?

**Comment by skulle** [comment]
The key resolution is only changed so far that it checks if a keyref points to a filtered map/topic or if a keydef is filtered out itself. Otherwise dead links could appear for example.

**Comment by jelovirt** [comment]
We already allow dead links, so IMO we should not touch key processing. It's up to the logger to filter out duplicate dead link messages.

**Comment by jelovirt** [comment]
An alternative would be to store unavailable file into into the `Job.FileInfo` metadata, so we know not to try to read them. But, it should be done in a separate PR.

**Comment by skulle** [comment]
> 
> 
> An alternative would be to store unavailable file into into the `Job.FileInfo` metadata, so we know not to try to read them. But, it should be done in a separate PR.

I am doing this currently in the map-branch-filter because i need the info later for the dead links. 

What do you mean by 'It's up to the logger to filter...'  Is that something a user of the toolkit should do with a custom class?

**Comment by jelovirt** [comment]
But why do you need information for the dead links when the issue is about supporting filtering out the whole topic? If filtering operations remove a file and create dead links, that's a different issue that should be part of a different PR. This PR for #3443 should only implement the support for filtering out the whole topic.

**Comment by skulle** [comment]
Initially, only the filtered topic was removed but then I still had my links and i thought I doesn't make sense. It is just that, there was no more profound need or so. 
Probably I can make a PR for only removing the topics and storing the info in the job file. Not sure if the first commit of mine will run through with all the tests. I will check that and make a new PR.

**Comment by jelovirt** [comment]
It's preferred to keep PR scope to a minimum, thus one PR for supporting filtering out topic content. When the whole topic has been filtered out, it should be removed from `Job` instead of marking it as deleted.

You can use this PR for the work and just rebase, instead of closing this and opening a new PR.

**Comment by skulle** [comment]
Of course, let me do the changes and I will come back to you... 

Thx for having a look at my PR.

**Comment by skulle** [comment]
Hi @jelovirt, the PR is refactored to remove filtered topics only. If you like you can have another look.

**Pull request closed without merging**